BUGFIX: Fixup #4679 constraints in node references

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Fixup #4679 constraints in node references

mhsdesign commented:
It seems though that we need a test for this ^^

kitsunet commented on Neos.ContentRepository.Core/Classes/NodeType/ConstraintCheck.php:
```diff
@@ -14,6 +14,15 @@
     private function __construct(
         private array $constraints
     ) {
+        foreach ($this->constraints as $constraint => $value) {
```
Shouldn't we build a `NodeTypes` collection similar to `Projections` and use that as argument here?

mhsdesign replied:
idk its your code ^^

mhsdesign replied:
i would argue the `ConstraintCheck` is already this collection

kitsunet replied:
Right, probably true, I guess I expected a "fromArray" or somthing.

mhsdesign committed:
TASK Adjust tests for NodeReferencing constraints

Now the tests are adjusted to test more real world behaviour and would have caught this regression: https://github.com/neos/neos-development-collection/pull/4798
